linkerd/app/outbound/src/protocol.rs

```diff
@@ -0,0 +1,142 @@
+use crate::{http, Outbound};
+pub use linkerd_app_core::proxy::api_resolve::ConcreteAddr;
+use linkerd_app_core::{detect, io, svc, Error, Infallible};
+use std::{fmt::Debug, hash::Hash};
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Http<T> {
+    version: http::Version,
+    parent: T,
+}
+
+/// Parameter type indicating how the proxy should handle a connection.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Protocol {
+    Http1,
+    Http2,
+    Detect,
+    Opaque,
+}
+
+// === impl Outbound ===
+
+impl<N> Outbound<N> {
+    /// Builds a stack that handles protocol detection as well as routing and
+    /// load balancing for a single logical destination.
+    ///
+    /// The inner stack is built once for each `T` target when the protocol is
+    /// known. When `Protocol::Detect` is used, the inner stack is built for
+    /// each connection.
+    pub fn push_protocol<T, I, H, HSvc, NSvc>(self, http: H) -> Outbound<svc::ArcNewTcp<T, I>>
+    where
+        // Target type indicating whether detection should be skipped.
+        T: svc::Param<Protocol>,
+        T: Eq + Hash + Clone + Debug + Send + Sync + 'static,
+        // Server-side socket.
+        I: io::AsyncRead + io::AsyncWrite + io::PeerAddr,
+        I: Debug + Send + Sync + Unpin + 'static,
+        // HTTP request stack.
+        H: svc::NewService<Http<T>, Service = HSvc>,
+        H: Clone + Send + Sync + Unpin + 'static,
+        HSvc: svc::Service<
+            http::Request<http::BoxBody>,
+            Response = http::Response<http::BoxBody>,
+            Error = Error,
+        >,
+        HSvc: Clone + Send + Sync + Unpin + 'static,
+        HSvc::Future: Send,
+        // Opaque connection stack.
+        N: svc::NewService<T, Service = NSvc>,
+        N: Clone + Send + Sync + Unpin + 'static,
+        NSvc: svc::Service<io::EitherIo<I, io::PrefixedIo<I>>, Response = (), Error = Error>,
+        NSvc: Clone + Send + Sync + Unpin + 'static,
+        NSvc::Future: Send,
+    {
+        let opaq = self.clone().into_stack();
+
+        let http = self.with_stack(http).map_stack(|config, rt, stk| {
+            stk.push_on_service(http::BoxRequest::layer())
+                .unlift_new()
+                .push(http::NewServeHttp::layer(
+                    config.proxy.server.h2_settings,
+                    rt.drain.clone(),
+                ))
+        });
+
+        // The inner stacks are built for each connection and not cached.
```

Comment:
tiny nit: maybe this comment should move to the top of the function, because it also applies to `opaq` and `http`?